File: solutions/problem_0401_binary_watch/solution.cpp
/*
 * Problem 401: Binary Watch
 * ========================
 * Difficulty: Easy
 * Tags: Backtracking, Bit Manipulation
 * Pattern: Backtracking
 *
 * Time Complexity:  O(k^n) or O(n!)
 * Space Complexity: O(n)
 */

#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<string> readBinaryWatch(int turnedOn) {
        // Backtracking - O(2^n) or O(n!) time
        vector<vector<int>> result;
        vector<int> path;
        function<void(int)> backtrack = [&](int start) {
            result.push_back(path);
            for (int i = start; i < (int)turnedOn.size(); i++) {
                path.push_back(turnedOn[i]);
                backtrack(i + 1);
                path.pop_back();
            }
        };
        backtrack(0);
        return result;
    }
};

File: solutions/problem_0401_binary_watch/solution.py
"""
Problem 401: Binary Watch
========================
Difficulty: Easy
Tags: Backtracking, Bit Manipulation
Pattern: Backtracking

Time Complexity:  O(k^n) or O(n!)
Space Complexity: O(n)
"""

class Solution:
    def readBinaryWatch(self, turnedOn: int) -> List[str]:
        # Backtracking - O(2^n) or O(n!) time
        result = []
        
        def backtrack(path, start):
            result.append(path[:])
            for i in range(start, len(turnedOn)):
                path.append(turnedOn[i])
                backtrack(path, i + 1)
                path.pop()
        
        backtrack([], 0)
        return result
